actionpack/lib/action_dispatch/testing/request_encoder.rb

```diff
@@ -9,6 +11,9 @@ def encode_params(params); params; end
       def response_parser; -> body { body }; end
     end
 
+    # :nodoc:
+    HTMLResponseParser = defined?(::Nokogiri::HTML5) ? ::Nokogiri::HTML5 : ::Nokogiri::HTML
```

Comment:
@flavorjones is this along the lines of what you had in mind?

@rafaelfranca will the `# :nodoc:` marker affect this constant declaration? Would it be better to define this as a class instance variable like `@html_response_parser`?

Reply:
:+1: This logic is right, thanks.

Though the test now fails with the HTML4 parser because it's asserting too narrowly on the serialized string, which differs between HTML4 and HTML5:

```
Failure:
TestResponseTest#test_response_parsing [/home/flavorjones/code/oss/rails/actionpack/test/dispatch/test_response_test.rb:31]:
--- expected
+++ actual
@@ -1 +1,6 @@
-"<html><head></head><body></body></html>"
+"<!DOCTYPE html PUBLIC \"-//W3C//DTD HTML 4.0 Transitional//EN\" \"http://www.w3.org/TR/REC-html40/loose.dtd\">
+<html>
+<head><meta http-equiv=\"Content-Type\" content=\"text/html; charset=UTF-8\"></head>
+<body></body>
+</html>
+"
```

What do you think about changing this to assert on a) the type of the object, and b) the structure of the document? Like this:

```patch
diff --git a/actionpack/test/dispatch/test_response_test.rb b/actionpack/test/dispatch/test_response_test.rb
index 575ac3f8..d43f2ef9 100644
--- a/actionpack/test/dispatch/test_response_test.rb
+++ b/actionpack/test/dispatch/test_response_test.rb
@@ -28,6 +28,8 @@ def assert_response_code_range(range, predicate)
     assert_equal({ "foo" => "fighters" }, response.parsed_body)
 
     response = ActionDispatch::TestResponse.create(200, { "Content-Type" => "text/html" }, "<html><head></head><body></body></html>")
-    assert_equal("<html><head></head><body></body></html>", response.parsed_body.to_html)
+    assert_kind_of(Nokogiri::XML::Document, response.parsed_body)
+    assert(response.parsed_body.at_css("/html/head"))
+    assert(response.parsed_body.at_css("/html/body"))
   end
 end
```

Reply:
Also, if it's important, you should test with some actual document content. By default an HTML5 document will parse an empty string with a `head` and `body` tag:

```ruby
Nokogiri::HTML5::Document.parse("").to_html
# => "<html><head></head><body></body></html>"
```

Maybe also just jam a `<div>content</div>` in there and then also assert on the presence of `response.parsed_body.at_xpath("/html/body/div")`.

Reply:
Done. Thank you!